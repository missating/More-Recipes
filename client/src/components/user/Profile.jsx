import React from 'react';
import Gravatar from 'react-gravatar';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';


// actions
import getUserProfile from '../../actions/userProfile';
import editProfile from '../../actions/editUserProfile';


/**
 * @description Creates Profile component
 *
 * @class Profile
 *
 * @extends {React.Component}
 */
export class Profile extends React.Component {
  /**
   * @description Creates an instance of Profile.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof Profile
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      fullname: '',
      username: '',
      email: '',
      joined: '',
      disabled: true,
    };
    this.onChange = this.onChange.bind(this);
    this.onEdit = this.onEdit.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.cancelEdit = this.cancelEdit.bind(this);
  }


  /**
  * @description Gets user details
  *
  * @method
  *
  * @memberof Profile
  *
  * @returns {void}
  */
  componentDidMount() {
    this.props.getUserDetails();
  }


  /**
   * @description Set state
   *
   * @param {object} nextProps
   *
   * @memberof Profile
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    const {
      fullname, username, email, joined,
    } = nextProps.userProfile;
    this.setState({
      fullname, username, email, joined,
    });
  }


  /**
   * @description Bind the value of the inputs to state
   *
   * @method onChange
   *
   * @param {object} event
   *
   * @memberof Profile
   *
   * @returns {void}
  */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }


  /**
   * @description Sets state to enable
   *
   * @memberof Profile
   *
   * @returns {void}
   */
  onEdit() {
    this.setState({ disabled: !this.state.disabled });
  }


  /**
   * @description Submit the form
   *
   * @method onSubmit
   *
   * @param {object} event
   *
   * @memberof Profile
   *
   * @returns {void}
   */
  onSubmit(event) {
    event.preventDefault();
    const { fullname, username } = this.state;
    this.props.updateProfile({
      fullname, username
    })
      .then(() => {
        this.setState({
          disabled: true
        });
      });
  }


  /**
   * @description Sets state to disable
   *
   * @memberof Profile
   *
   * @returns {void}
   */
  cancelEdit() {
    this.setState({
      disabled: true
    });
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof Profile
   *
   * @returns {void}
   *
   */
  render() {
    const { isFetching } = this.props;

    if (isFetching) {
      return (
        <div className="loader-container">
          <div className="loader" />
        </div>
      );
    }

    return (
      <div className="container profile" id="recipes">
        {
          !this.props.authenticated &&
          <Redirect to="/" />
        }

        <h3 className="text-center">Profile</h3>
        <hr />

        <div className="row">
          <div className="col-md-6 col-sm-12">
            <Gravatar
              className="rounded mx-auto d-block mt-3"
              width="200"
              height="200"
              email={this.state.email}
            />
          </div>

          <div className="col-md-6 col-sm-12">

            <div className="form-group form-width">
              <h5 className="pt-3"> Full Name </h5>
              <input
                type="text"
                className="form-control"
                name="fullname"
                value={this.state.fullname}
                onChange={this.onChange}
                disabled={this.state.disabled}
              />
            </div>
            <form>
              <div className="form-group form-width">
                <h5 className="pt-3"> User Name </h5>
                <input
                  type="text"
                  className="form-control"
                  id="usernameField"
                  name="username"
                  value={this.state.username}
                  onChange={this.onChange}
                  disabled={this.state.disabled}
                />
              </div>

              <div className="form-group form-width">
                <h5 className="pt-3"> Email </h5>
                <input
                  type="email"
                  className="form-control"
                  name="email"
                  value={this.state.email}
                  onChange={this.onChange}
                  disabled
                />
              </div>

              <div className="form-group form-width">
                <h5 className="pt-3"> Joined </h5>
                <input
                  type="text"
                  className="form-control"
                  name="joined"
                  value={this.state.joined}
                  onChange={this.onChange}
                  disabled
                />

                <div
                  className="btn-group"
                  role="group"
                  aria-label="Basic example"
                >
                  <button
                    type="button"
                    id="profileEdit"
                    className="btn btn-secondary"
                    onClick={this.onEdit}
                    disabled={!this.state.disabled}
                  >
                    Edit
                  </button>
                  <button
                    type="button"
                    id="profileSubmit"
                    className="btn btn-secondary"
                    disabled={this.state.disabled}
                    onClick={this.onSubmit}
                  >
                    Submit
                  </button>
                  <button
                    type="button"
                    id="profileCancel"
                    className="btn btn-secondary"
                    onClick={this.cancelEdit}
                    disabled={this.state.disabled}
                  >
                    Cancel
                  </button>
                </div>
              </div>
            </form>

          </div>

        </div>
      </div >
    );
  }
}

Profile.propTypes = {
  userProfile: PropTypes.shape({}).isRequired,
  getUserDetails: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  updateProfile: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  userProfile: state.userProfile,
  authenticated: state.auth.isAuthenticated,
  isFetching: state.isFetching
});

const mapDispatchToProps = dispatch => ({
  getUserDetails: () => dispatch(getUserProfile()),
  updateProfile: userDetails => dispatch(editProfile(userDetails))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(Profile);
export { ConnectedComponent as ConnectedProfile };
