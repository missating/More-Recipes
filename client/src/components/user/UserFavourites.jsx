import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Redirect } from 'react-router-dom';
import { confirmAlert } from 'react-confirm-alert';
import 'react-confirm-alert/src/react-confirm-alert.css';

// components
import RecipeCard from '../cards/RecipeCard';

// actions
import getUserFavourites from '../../actions/getUserFavourites';
import { removeUserFavourite } from '../../actions/addFavourite';

/**
 * @description Creates UserFavourites component
 *
 * @class UserFavourites
 *
 * @extends {React.Component}
 */
export class UserFavourites extends React.Component {
  /**
   * @description Creates an instance of UserFavourites.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
    };
    this.removeFromFavourite = this.removeFromFavourite.bind(this);
    this.onPageChange = this.onPageChange.bind(this);
  }


  /**
   * @description Gets user's favourites
   *
   * @method
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  componentDidMount() {
    this.props.favourites();
  }


  /**
   * @description Set state
   *
   * @param {object} nextProps
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.userFavourites !== nextProps.userFavourites) {
      this.setState({
        userFavourites: nextProps.userFavourites,
        loading: false
      });
    }
  }


  /**
   * @description pagination for user favourites
   *
   * @param {number} current
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  onPageChange(current) {
    const selected = current.selected + 1;
    this.props.history.push(`/users/favourites?page=${selected}`);
    this.props.favourites(selected);
  }

  /**
   * @description removes a user's favourites
   *
   * @param {number} id
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  removeFromFavourite(id) {
    confirmAlert({
      title: '',
      message: 'Are you sure you want to remove this recipe ?',
      confirmLabel: 'Yes, remove!',
      cancelLabel: 'Cancel',
      onConfirm: () =>
        this.props.removeFavourite(id),
    });
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof UserFavourites
   *
   * @returns {void}
   */
  render() {
    const { loading } = this.state;

    if (loading) {
      return (
        <div className="loader-container">
          <div className="loader" />
        </div>
      );
    }

    const { pages } = this.props.pagination;

    const userFavourites =
      (this.state.userFavourites) ?
        this.state.userFavourites : [];

    let userFavouritesError;
    if (userFavourites.length === 0) {
      userFavouritesError = (
        <span className="help-block">
          You have no recipes added as favourites
        </span>
      );
    }


    const userFavouritesList = userFavourites.map((favourites, i) => (
      <div className="col-md-4" key={`favourites${i + 1}`}>
        <RecipeCard
          recipeImage={favourites.Recipe.recipeImage}
          name={favourites.Recipe.name}
          description={favourites.Recipe.description}
          id={favourites.recipeId}
          button
          onButtonClick={this.removeFromFavourite}
        />
      </div>
    ));

    if (userFavourites.length) {
      return (
        <section className="container" id="recipes">
          <div>
            {
              !this.props.authenticated &&
              <Redirect to="/" />
            }

            <h3 className="text-center">My Favourite Recipes</h3>
            <hr />
            <div className="row mh-500">
              {userFavouritesList}
            </div>
            <div className="container">
              <ReactPaginate
                pageCount={pages}
                pageRangeDisplayed={5}
                marginPagesDisplayed={3}
                previousLabel="Previous"
                nextLabel="Next"
                breakClassName="text-center"
                initialPage={0}
                containerClassName="container pagination justify-content-center"
                pageClassName="page-item"
                pageLinkClassName="page-link"
                activeClassName="page-item active"
                previousClassName="page-item"
                nextClassName="page-item"
                nextLinkClassName="page-link"
                previousLinkClassName="page-link"
                onPageChange={this.onPageChange}
              />
            </div>

          </div>
        </section>

      );
    }
    return (
      <section className="container" id="recipes">
        <div>
          {
            !this.props.authenticated &&
            <Redirect to="/" />
          }

          <h3 className="text-center">My Favourite Recipes</h3>
          <hr />
          <h4 className="text-center m-5"> {userFavouritesError} </h4>
        </div>
      </section>
    );
  }
}

UserFavourites.propTypes = {
  favourites: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  pagination: PropTypes.shape({
    pages: PropTypes.number
  }),
  userFavourites: PropTypes.arrayOf(PropTypes.shape({})),
  removeFavourite: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
};

UserFavourites.defaultProps = {
  pagination: {
    pages: 1
  },
  userFavourites: []
};


const mapStateToProps = state => ({
  userFavourites: state.userFavourites.favourites,
  authenticated: state.auth.isAuthenticated,
  pagination: state.pagination,
  isFetching: state.isFetching
});

const mapDispatchToProps = dispatch => ({
  favourites: page => dispatch(getUserFavourites(page)),
  removeFavourite: recipedId => dispatch(removeUserFavourite(recipedId))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(UserFavourites);
export { ConnectedComponent as ConnectedUserFavourites };
