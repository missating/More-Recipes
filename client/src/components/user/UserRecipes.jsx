import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Redirect } from 'react-router-dom';

// components
import { ConnectedUserRecipeCard } from '../cards/UserRecipeCard';

// actions
import getUserRecipes from '../../actions/getUserRecipes';

/**
 * @description Creates UserRecipes component
 *
 * @class UserRecipes
 *
 * @extends {React.Component}
 */
export class UserRecipes extends React.Component {
  /**
   * @description Creates an instance of UserFavourites.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof UserRecipes
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      loading: true
    };
    this.onPageChange = this.onPageChange.bind(this);
  }


  /**
   * @description Gets user's recipes
   *
   * @method
   *
   * @memberof UserRecipes
   *
   * @returns {void}
   */
  componentDidMount() {
    this.props.recipes();
  }

  /**
   * @description Set state
   *
   * @param {object} nextProps
   *
   * @memberof UserRecipes
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.userRecipes !== nextProps.userRecipes) {
      this.setState({
        userRecipes: nextProps.userRecipes,
        loading: false
      });
    }
  }


  /**
   * @description pagination for user recipes
   *
   * @param {number} current
   *
   * @memberof UserRecipes
   *
   * @returns {void}
   */
  onPageChange(current) {
    const selected = current.selected + 1;
    this.props.history.push(`/users/recipes?page=${selected}`);
    this.props.recipes(selected);
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof UserRecipes
   *
   * @returns {void}
   */
  render() {
    const { loading } = this.state;

    if (loading) {
      return (
        <div className="loader-container">
          <div className="loader" />
        </div>
      );
    }

    const { pages } = this.props.pagination;

    const userRecipes =
      (this.state.userRecipes) ?
        this.state.userRecipes : [];

    let userRecipeError;
    if (userRecipes.length === 0) {
      userRecipeError = (
        <span className="help-block">
          You currently have no recipes
        </span>
      );
    }

    const userRecipesList = userRecipes.map((recipe, i) => (
      <div className="col-md-6 col-lg-4 col-sm-6 p-0" key={`recipe${i + 1}`}>
        <ConnectedUserRecipeCard
          {...recipe}
        />
      </div>
    ));
    if (userRecipes.length) {
      return (
        <section className="container p-0" id="recipes">
          <div>
            {
              !this.props.authenticated &&
              <Redirect to="/" />
            }

            <h3 className="text-center">My Recipes</h3>
            <hr />
            <div className="row mh-500">
              {userRecipesList}
            </div>

            <div className="container">
              <ReactPaginate
                pageCount={pages}
                pageRangeDisplayed={5}
                marginPagesDisplayed={3}
                previousLabel="Previous"
                nextLabel="Next"
                breakClassName="text-center"
                initialPage={0}
                containerClassName="container pagination justify-content-center"
                pageClassName="page-item"
                pageLinkClassName="page-link"
                activeClassName="page-item active"
                previousClassName="page-item"
                nextClassName="page-item"
                nextLinkClassName="page-link"
                previousLinkClassName="page-link"
                onPageChange={this.onPageChange}
              />
            </div>

          </div>
        </section>
      );
    }
    return (
      <section className="section" id="recipes">
        <div>
          {
            !this.props.authenticated &&
            <Redirect to="/" />
          }

          <h3 className="text-center">My Recipes</h3>
          <hr />
          <h4 className="text-center m-5"> {userRecipeError} </h4>
        </div>
      </section>
    );
  }
}

UserRecipes.propTypes = {
  recipes: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  pagination: PropTypes.shape({
    pages: PropTypes.number
  }),
  userRecipes: PropTypes.arrayOf(PropTypes.shape({})),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
};

UserRecipes.defaultProps = {
  pagination: {
    pages: 1
  },
  userRecipes: []
};

const mapStateToProps = state => ({
  userRecipes: state.userRecipes.recipes,
  authenticated: state.auth.isAuthenticated,
  pagination: state.pagination,
  isFetching: state.isFetching
});

const mapDispatchToProps = dispatch => ({
  recipes: page => dispatch(getUserRecipes(page))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(UserRecipes);
export { ConnectedComponent as ConnectedUserRecipes };

