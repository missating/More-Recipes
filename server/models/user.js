import bcrypt from 'bcrypt';

export default (sequelize, DataTypes) => {
  const User = sequelize.define(
    'User', {
      fullname: {
        type: DataTypes.STRING,
        allownull: false
      },
      username: {
        type: DataTypes.STRING,
        allownull: false
      },
      email: {
        type: DataTypes.STRING,
        unique: true,
        allowNull: false,
        validate: {
          isEmail: true
        }
      },
      password: {
        type: DataTypes.STRING,
        allowNull: false
      }
    },
    {
      hooks: {
        beforeCreate: (user) => {
          user.password = bcrypt.hashSync(user.password, 10);
        }
      }
    }
  );

  User.associate = (models) => {
    // associations defined here
    User.hasMany(models.Recipe, { foreignKey: 'userId', onDelete: 'CASCADE' });
    User.hasMany(models.Review, { foreignKey: 'userId' });
    User.hasMany(models.Vote, { foreignKey: 'userId', onDelete: 'CASCADE' });
    User.hasMany(
      models.Favourite,
      { foreignKey: 'userId', onDelete: 'CASCADE' }
    );
  };
  return User;
};
