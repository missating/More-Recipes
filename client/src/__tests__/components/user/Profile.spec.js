import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import { Profile } from '../../../components/user/Profile';

describe('Profile component', () => {
  const props = {
    userProfile: {
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17'
    },
    getUserDetails: jest.fn(() => Promise.resolve()),
    authenticated: true,
    updateProfile: jest.fn(() => Promise.resolve()),
    isFetching: false
  };

  it('Should render', () => {
    const wrapper = shallow(<Profile
      {...props}
    />);
    expect(wrapper.find('form').exists).toBeTruthy();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });


  it('Should render loader when fetching', () => {
    const props2 = {
      userProfile: {
        fullname: 'test',
        username: 'test',
        email: 'test@test.com',
        joined: '3/4/17'
      },
      getUserDetails: jest.fn(() => Promise.resolve()),
      authenticated: true,
      updateProfile: jest.fn(() => Promise.resolve()),
      isFetching: true
    };

    const wrapper = shallow(<Profile
      {...props2}
    />);
    expect(wrapper.find('div').length).toEqual(2);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });


  it(
    'calls onChange method when the value of the new recipe input changes',
    () => {
      const spy = sinon.spy(Profile.prototype, 'onChange');
      const wrapper = shallow(<Profile
        {...props}
      />);
      wrapper.find('#usernameField').simulate('change', {
        target: { name: 'username', value: 'testy' }
      });
      expect(spy.called).toBeTruthy();
    }
  );

  it('Should submit form when submit button is clicked', () => {
    const spy = sinon.spy(Profile.prototype, 'onSubmit');
    const wrapper = shallow(<Profile
      {...props}
    />);
    wrapper.find('#profileSubmit').simulate(
      'click',
      { preventDefault: jest.fn() }
    );
    expect(spy.called).toBeTruthy();
  });


  it('Should enable state when edit button is clicked', () => {
    const spy = sinon.spy(Profile.prototype, 'onEdit');
    const wrapper = shallow(<Profile
      {...props}
    />);
    wrapper.instance().setState({
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17',
      disabled: true
    });
    wrapper.update();
    wrapper.find('#profileEdit').simulate('click');
    expect(wrapper.instance().state).toEqual({
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17',
      disabled: false
    });
    expect(spy.called).toBeTruthy();
  });


  it('Should disable state when cancel button is clicked', () => {
    const spy = sinon.spy(Profile.prototype, 'cancelEdit');
    const wrapper = shallow(<Profile
      {...props}
    />);
    wrapper.instance().setState({
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17',
      disabled: false
    });
    wrapper.update();
    wrapper.find('#profileCancel').simulate('click');
    expect(wrapper.instance().state).toEqual({
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17',
      disabled: true
    });
    expect(spy.called).toBeTruthy();
  });


  it('Should be called when component is passed in new props', () => {
    const wrapper = shallow(<Profile
      {...props}
    />);
    wrapper.instance().componentWillReceiveProps(props);
    expect(wrapper.instance().props.userProfile).toEqual(props.userProfile);
    wrapper.instance().setState({
      fullname: 'test',
      username: 'test',
      email: 'test@test.com',
      joined: '3/4/17',
    });
    wrapper.update();
  });
});

