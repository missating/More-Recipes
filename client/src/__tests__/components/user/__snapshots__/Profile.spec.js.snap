// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component Should render 1`] = `
<div
  className="container profile"
  id="recipes"
>
  <h3
    className="text-center"
  >
    Profile
  </h3>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-6 col-sm-12"
    >
      <Gravatar
        className="rounded mx-auto d-block mt-3"
        default="retro"
        email=""
        height="200"
        protocol="//"
        rating="g"
        size={50}
        width="200"
      />
    </div>
    <div
      className="col-md-6 col-sm-12"
    >
      <div
        className="form-group form-width"
      >
        <h5
          className="pt-3"
        >
           Full Name 
        </h5>
        <input
          className="form-control"
          disabled={true}
          name="fullname"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <form>
        <div
          className="form-group form-width"
        >
          <h5
            className="pt-3"
          >
             User Name 
          </h5>
          <input
            className="form-control"
            disabled={true}
            id="usernameField"
            name="username"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group form-width"
        >
          <h5
            className="pt-3"
          >
             Email 
          </h5>
          <input
            className="form-control"
            disabled={true}
            name="email"
            onChange={[Function]}
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group form-width"
        >
          <h5
            className="pt-3"
          >
             Joined 
          </h5>
          <input
            className="form-control"
            disabled={true}
            name="joined"
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            aria-label="Basic example"
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-secondary"
              disabled={false}
              id="profileEdit"
              onClick={[Function]}
              type="button"
            >
              Edit
            </button>
            <button
              className="btn btn-secondary"
              disabled={true}
              id="profileSubmit"
              onClick={[Function]}
              type="button"
            >
              Submit
            </button>
            <button
              className="btn btn-secondary"
              disabled={true}
              id="profileCancel"
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Profile component Should render loader when fetching 1`] = `
<div
  className="loader-container"
>
  <div
    className="loader"
  />
</div>
`;
